yt/frontends/swift/io.py

```diff
@@ -0,0 +1,168 @@
+            dt = ds.dtype.newbyteorder("N") # Native
+            if position_dtype is not None and dt < position_dtype:
+                # Sometimes positions are stored in double precision
+                # but smoothing lengths are stored in single precision.
+                # In these cases upcast smoothing length to double precision
+                # to avoid ValueErrors when we pass these arrays to Cython.
+                dt = position_dtype
+            hsml = np.empty(ds.shape, dtype=dt)
+            hsml[:] = ds
+            return hsml
+
+    def _read_particle_fields(self, chunks, ptf, selector):
+        # Now we have all the sizes, and we can allocate
+        data_files = set([])
+        for chunk in chunks:
+            for obj in chunk.objs:
+                data_files.update(obj.data_files)
+
+        for data_file in sorted(data_files, key=lambda x: x.filename):
+            si, ei = data_file.start, data_file.end
+            f = h5py.File(data_file.filename, "r")
+            for ptype, field_list in sorted(ptf.items()):
+                if data_file.total_particles[ptype] == 0:
+                    continue
+                g = f["/%s" % ptype]
+                coords = g["Coordinates"][si:ei].astype("float64")
+                if ptype == 'PartType0':
+                    hsmls = g["SmoothingLength"][si:ei].astype("float64")
+                else:
+                    hsmls = 0.0
+                mask = selector.select_points(
+                    coords[:,0], coords[:,1], coords[:,2], hsmls)
+                del coords
+                if mask is None:
+                    continue
+                for field in field_list:
+                    if field in ("Mass", "Masses"):
+                        data = g[self.ds._particle_mass_name][si:ei][mask, ...]
+                    else:
+                        data = g[field][si:ei][mask, ...]
+
+                    yield (ptype, field), data
+            f.close()
+
+    def _count_particles(self, data_file):
+        si, ei = data_file.start, data_file.end
+        f = h5py.File(data_file.filename, "r")
+        pcount = f["/Header"].attrs["NumPart_ThisFile"][:].astype("int")
+        f.close()
+        if None not in (si, ei):
+            np.clip(pcount - si, 0, ei - si, out=pcount)
+        npart = dict(("PartType%s" % (i), v) for i, v in enumerate(pcount))
+        return npart
+
+    def _identify_fields(self, data_file):
+        f = h5py.File(data_file.filename, "r")
+        fields = []
+        cname = self.ds._particle_coordinates_name  # Coordinates
+        mname = self.ds._particle_mass_name  # Coordinates
+
+        for key in f.keys():
+            if not key.startswith("PartType"):
+                continue
+
+            g = f[key]
+            if cname not in g:
+                continue
+
+            ptype = str(key)
+            for k in g.keys():
+                    kk = k
+                    # a little hack to deal with swift masses
```

Comment:
i don't see what the "hack" is here. Can you provide a little more detail? If it's not worth over-explaining I'd just delete the comment.